internal/blockchain/thresholdstate.go

```diff
@@ -338,99 +288,103 @@ func (b *BlockChain) nextThresholdState(prevNode *blockNode, deployment *deploym
+			voteParams := &deployment.deployment.Vote
+			choiceIdxShift := bits.TrailingZeros16(voteParams.Mask)
+
+			// At this point, the rule change is still being voted on, so
+			// iterate backwards through the confirmation window to count all of
+			// the votes in it.
+			var totalAbstainVotes, totalNonAbstainVotes uint32
+			choiceCounts := make([]uint32, len(voteParams.Choices))
 			countNode := prevNode
 			for i := int64(0); i < confirmationWindow; i++ {
-				c, err := checker.Condition(countNode, version)
-				if err != nil {
-					return newThresholdState(
-						ThresholdFailed, invalidChoice), err
-				}
+				for _, vote := range countNode.votes {
+					// Ignore votes that are not for the deployment version.
+					if vote.Version != deployment.version {
+						continue
+					}
 
-				// Create array first time around.
-				if len(counts) == 0 {
-					counts = make([]thresholdConditionTally, len(c))
-				}
+					// Ignore votes for invalid choices.
+					choiceIdx := (voteParams.Mask & vote.Bits) >> choiceIdxShift
+					if choiceIdx > uint16(len(voteParams.Choices)-1) {
+						continue
+					}
 
-				// Tally votes.
-				for k := range c {
-					counts[k].count += c[k].count
-					counts[k].isAbstain = c[k].isAbstain
-					counts[k].isNo = c[k].isNo
-					if c[k].isAbstain {
-						abstainVotes += c[k].count
+					choiceCounts[choiceIdx]++
+					if voteParams.Choices[choiceIdx].IsAbstain {
+						totalAbstainVotes++
 					} else {
-						totalVotes += c[k].count
+						totalNonAbstainVotes++
 					}
 				}
 
 				countNode = countNode.parent
 			}
 
-			// Determine if we have reached quorum.
-			totalNonAbstainVotes := uint32(0)
-			for _, v := range counts {
-				if v.isAbstain && !v.isNo {
-					continue
-				}
-				totalNonAbstainVotes += v.count
-			}
-			if totalNonAbstainVotes < checker.RuleChangeActivationQuorum() {
+			// The state remains in the started state when the quorum has not
+			// been met since the vote requires a quorum for a concrete result.
+			if totalNonAbstainVotes < b.chainParams.RuleChangeActivationQuorum {
 				break
 			}
 
-			// The state is locked in if the number of blocks in the
-			// period that voted for the rule change meets the
+			// The state moves to either locked in or failed with a specific
+			// non-abstaining choice when any of them meets the majority
 			// activation threshold.
-			for k, v := range counts {
-				// We require at least 10% quorum on all votes.
-				if v.count < checker.RuleChangeActivationThreshold(totalVotes) {
-					continue
+			ruleChangeActivationThreshold := totalNonAbstainVotes *
+				b.chainParams.RuleChangeActivationMultiplier /
+				b.chainParams.RuleChangeActivationDivisor
+		nextChoice:
+			for choiceIdx := range voteParams.Choices {
+				// Abstain never changes the state.
+				choice := &voteParams.Choices[choiceIdx]
+				if choice.IsAbstain {
+					if choice.IsNo {
+						return newThresholdState(ThresholdFailed, invalidChoice),
```

Comment:
Isn't this redundant? We have already checked that no choices are both abstain and no in `validateDeploymentChoices`?

Reply:
Good call.  I had this here (and similar in other places) originally and decided to do the validation once when the chain is initialized instead and forgot to remove this one.  Updated.